processing/src/main/java/io/druid/segment/incremental/OffHeapIncrementalIndex.java

```diff
@@ -0,0 +1,528 @@
+
+        for (int i = 0; i < metrics.length; i++) {
+          getAggs()[i].init(aggBuffer, bufferOffset + aggOffsetInBuffer[i]);
+        }
+
+        // Last ditch sanity checks
+        if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
+          throw new IndexSizeExceededException("Maximum number of rows [%d] reached", maxRowCount);
+        }
+
+        final Integer rowIndex = indexIncrement.getAndIncrement();
+        final Integer prev = facts.putIfAbsent(key, rowIndex);
+        if (null == prev) {
+          numEntries.incrementAndGet();
+          indexAndOffsets.add(new int[]{bufferIndex, bufferOffset});
+        } else {
+          throw new ISE("WTF! we are in sychronized block.");
+        }
+      }
+    }
+
+    rowContainer.set(row);
+
+    for (int i = 0; i < metrics.length; i++) {
+      final BufferAggregator agg = getAggs()[i];
+
+      synchronized (agg) {
+        agg.aggregate(aggBuffer, bufferOffset + aggOffsetInBuffer[i]);
+      }
+    }
+    rowContainer.set(null);
+    return numEntries.get();
+  }
+
+  @Override
+  public boolean canAppendRow()
+  {
+    final boolean canAdd = size() < maxRowCount;
+    if (!canAdd) {
+      outOfRowsReason = String.format("Maximum number of rows [%d] reached", maxRowCount);
+    }
+    return canAdd;
+  }
+
+  @Override
+  public String getOutOfRowsReason()
+  {
+    return outOfRowsReason;
+  }
+
+  @Override
+  protected BufferAggregator[] getAggsForRow(int rowOffset)
+  {
+    return getAggs();
+  }
+
+  @Override
+  protected Object getAggVal(BufferAggregator agg, int rowOffset, int aggPosition)
+  {
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.get(bb, indexAndOffset[1] + aggOffsetInBuffer[aggPosition]);
+  }
+
+  @Override
+  public float getMetricFloatValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.getFloat(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  public long getMetricLongValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.getLong(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  @Override
+  public Object getMetricObjectValue(int rowOffset, int aggOffset)
+  {
+    BufferAggregator agg = getAggs()[aggOffset];
+    int[] indexAndOffset = indexAndOffsets.get(rowOffset);
+    ByteBuffer bb = aggBuffers.get(indexAndOffset[0]).get();
+    return agg.get(bb, indexAndOffset[1] + aggOffsetInBuffer[aggOffset]);
+  }
+
+  /**
+   * NOTE: This is NOT thread-safe with add... so make sure all the adding is DONE before closing
+   */
+  @Override
+  public void close()
```

Comment:
Could this simply have a closer which handles the various closing and clearing steps instead of calling each one separately? that way any errors are guaranteed handled and the rest of the stuff still closed.